runtime/src/gpu/amd/gpu-amd.c

```diff
@@ -293,8 +293,8 @@ static void chpl_gpu_launch_kernel_help(int ln,
     }
   }
 
-  // TODO: this should use chpl_mem_free
-  chpl_free(kernel_params);
+  chpl_mem_free(kernel_params, fn, ln);
+  chpl_mem_free(was_memory_dynamically_allocated_for_kernel_param, fn, ln);
```

Comment:
Thanks for adding this missing `free`.

Note that `chpl_mem*` interface is the one where memory tracking and internal bookkeeping occurs. For `chpl_mem_free` to work correctly, the corresponding allocation calls should also use `chpl_mem*` interface instead of `chpl_*` interface. As-is, probably `--memTrack` will report a double free (or freeing something that's not been allocated) if it is sensitive for things like that. That's because the allocations are never captured by the memory tracking interface in the runtime.

I think it is worthwhile to try to use `chpl_mem`, but not the most important thing, so I was pushing it off whenever I see those `TODO`s. What needs to be done:

- `chpl_malloc` calls should turn into `chpl_mem_alloc`.
- `chpl_mem_alloc` takes a mem descriptor, so, ideally, we should add a `GPU_KERNEL_PARAM_BUFF` or something here: https://github.com/chapel-lang/chapel/blob/main/runtime/include/chpl-mem-desc.h#L98-L99 and use that.
- trivially, `ln` and `fn` should be threaded into that function call

Reply:
Ok, I've updated to use `chpl_mem_alloc` and added some additional memory descriptors.

Though, I'm a little confused on what the cutoff point should be for what we want tracked with `--memTrack` and what we don't (should all allocations in the runtime be done through `chpl_mem*`?)

Anyway, could you take another look and review?

Reply:
> Though, I'm a little confused on what the cutoff point should be for what we want tracked with --memTrack and what we don't (should all allocations in the runtime be done through chpl_mem*?)

Yeah, I don't know either. But a grep on `chpl_malloc` shows that the GPU code is the main user of that interface, so we should probably stick to what the rest of the runtime does here, which seems to point to a "yes" for your question.